#! /bin/sh /usr/share/dpatch/dpatch-run
## r188.dpatch by Andrey Rahmatullin <wrar@wrar.name>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Add some new DWARF definitions from the upstream r188.

@DPATCH@

Index: prelink/src/dwarf2.c
===================================================================
--- prelink/src/dwarf2.c	(revision 187)
+++ prelink/src/dwarf2.c	(revision 188)
@@ -403,8 +403,8 @@
 	  break;
 	case DW_OP_implicit_value:
 	  {
-	    uint32_t len = read_uleb128 (ptr);
-	    ptr += len;
+	    uint32_t leni = read_uleb128 (ptr);
+	    ptr += leni;
 	  }
 	  break;
 	case DW_OP_GNU_implicit_pointer:
@@ -420,6 +420,20 @@
 	    ptr += 4;
 	  read_uleb128 (ptr);
 	  break;
+        case DW_OP_GNU_entry_value:
+	  {
+	    uint32_t leni = read_uleb128 (ptr);
+	    if ((end - ptr) < leni)
+	      {
+		error (0, 0, "%s: DWARF DW_OP_GNU_entry_value with too large"
+		       " length", dso->filename);
+		return 1;
+	      }
+	    if (adjust_location_list (dso, cu, ptr, leni, start, adjust))
+	      return 1;
+	    ptr += leni;
+	  }
+	  break;
 	default:
 	  error (0, 0, "%s: Unknown DWARF DW_OP_%d", dso->filename, op);
 	  return 1;
@@ -706,6 +720,10 @@
 		case DW_AT_location:
 		case DW_AT_data_member_location:
 		case DW_AT_vtable_elem_location:
+		case DW_AT_GNU_call_site_value:
+		case DW_AT_GNU_call_site_data_value:
+		case DW_AT_GNU_call_site_target:
+		case DW_AT_GNU_call_site_target_clobbered:
 		  if (adjust_location_list (dso, cu, ptr, len, start, adjust))
 		    return NULL;
 		  break;
Index: prelink/src/dwarf2.h
===================================================================
--- prelink/src/dwarf2.h	(revision 187)
+++ prelink/src/dwarf2.h	(revision 188)
@@ -82,6 +82,11 @@
 #define DW_TAG_class_template		0x4103
 #define DW_TAG_GNU_BINCL		0x4104
 #define DW_TAG_GNU_EINCL		0x4105
+#define DW_TAG_GNU_template_template_param 0x4106
+#define DW_TAG_GNU_template_parameter_pack 0x4107
+#define DW_TAG_GNU_formal_parameter_pack 0x4108
+#define DW_TAG_GNU_call_site		0x4109
+#define DW_TAG_GNU_call_site_parameter	0x410a
 #define DW_TAG_lo_user			0x4080
 #define DW_TAG_hi_user			0xffff
 
@@ -228,6 +233,24 @@
 #define DW_AT_src_coords		0x2104
 #define DW_AT_body_begin		0x2105
 #define DW_AT_body_end			0x2106
+#define DW_AT_GNU_vector		0x2107
+#define DW_AT_GNU_guarded_by		0x2108
+#define DW_AT_GNU_pt_guarded_by		0x2109
+#define DW_AT_GNU_guarded		0x210a
+#define DW_AT_GNU_pt_guarded		0x210b
+#define DW_AT_GNU_locks_excluded	0x210c
+#define DW_AT_GNU_exclusive_locks_required 0x210d
+#define DW_AT_GNU_shared_locks_required	0x210e
+#define DW_AT_GNU_odr_signature		0x210f
+#define DW_AT_GNU_template_name		0x2110
+#define DW_AT_GNU_call_site_value	0x2111
+#define DW_AT_GNU_call_site_data_value	0x2112
+#define DW_AT_GNU_call_site_target	0x2113
+#define DW_AT_GNU_call_site_target_clobbered 0x2114
+#define DW_AT_GNU_tail_call		0x2115
+#define DW_AT_GNU_all_tail_call_sites	0x2116
+#define DW_AT_GNU_all_call_sites	0x2117
+#define DW_AT_GNU_all_source_call_sites	0x2118
 #define DW_AT_lo_user			0x2000
 #define DW_AT_hi_user			0x3ff0
 
@@ -389,6 +412,7 @@
 #define DW_OP_GNU_uninit		0xf0
 #define DW_OP_GNU_encoded_addr		0xf1
 #define DW_OP_GNU_implicit_pointer	0xf2
+#define DW_OP_GNU_entry_value		0xf3
 #define DW_OP_lo_user			0xe0
 #define DW_OP_hi_user			0xff
 
